File: typescript/bob/bob.ts
class Bob {
  hey(sentence: string): string {
    if (this.forcefulQuestion(sentence)) {
      return "Calm down, I know what I'm doing!";
    } else if (this.question(sentence)) {
      return 'Sure.';
    } else if (this.yelling(sentence)) {
      return 'Whoa, chill out!';
    } else if (this.silence(sentence)) {
      return 'Fine. Be that way!';
    } else {
      return 'Whatever.';
    }
  }

  private forcefulQuestion(sentence: string): boolean {
    return this.yelling(sentence) && this.question(sentence);
  }

  private yelling(sentence: string): boolean {
    return sentence.toUpperCase() === sentence && this.containsWord(sentence);
  }

  private question(sentence: string): boolean {
    return sentence.trim().endsWith('?');
  }

  private silence(sentence: string): boolean {
    return sentence.trim() === '';
  }

  private containsWord(sentence: string): boolean {
    return /[a-zA-Z]+/.test(sentence);
  }
}

export default Bob;

File: go/bob/bob.go
// This is a "stub" file.  It's a little start on your solution.
// It's not a complete solution though; you have to write some code.

// Package bob should have a package comment that summarizes what it's about.
// https://golang.org/doc/effective_go.html#commentary
package bob

import (
	"strings"
	"unicode"
)

// Hey should have a comment documenting it.
func Hey(remark string) string {
	remark = strings.TrimSpace(remark)
	if remark == "" {
		return "Fine. Be that way!"
	}

	switch {
	case strings.HasSuffix(remark, "?") && isUpper(remark):
		return "Calm down, I know what I'm doing!"
	case strings.HasSuffix(remark, "?"):
		return "Sure."
	case isUpper(remark):
		return "Whoa, chill out!"
	default:
		return "Whatever."
	}
}

func isUpper(remark string) bool {
	containsLetter := false
	for _, r := range remark {
		if unicode.IsLetter(r) {
			containsLetter = true
			if !unicode.IsUpper(r) {
				return false
			}
		}
	}
	return containsLetter
}
